### DISCUSSION TITLE
How to use NameValueListBase in CSLA 6.2.2

### DISCUSSION THREAD

**Discussion by CROprogrammer** [MarimerLLC/csla#3326] [Questions]
Hi there, is there any example that shows how to use NameValueListBase without factory methods in it?

**Comment by rockfordlhotka** [comment]
Like this you mean?

```c#
  [Serializable]
  public class NVList : NameValueListBase<string, string>
  {
    [Fetch]
    private void Fetch([Inject]DataAccess.IListDal dal)
    {
      using (LoadListMode)
      {
        // load data here
      }
    }
  }
```

And then you get it via the data portal:

```c#
  var dp = ApplicationContext.GetRequiredService<IDataPortal<ListInfo>>();
  var nvList = await dp.FetchAsync();
```

**Reply by CROprogrammer** [reply]
Should I still use private member variable in such classes. For example in ProjectTracker in RoleList class => private static RoleList CachedList;
And should I still use static methods to invalidate cache, getting default role, etc... ?

**Reply by rockfordlhotka** [reply]
You really can't use a static method with dependency injection. The alternative is to create a service (class) that manages your cache. Or use REDIS if you are in a server-side scenario.

For a smart client, you can either a CacheManager type something like this:

```c#
// in program.cs
builder.Services.AddScoped<CacheManager>();

// cache manager service
public class CacheManager
{
    public CacheManager(ApplicationContext applicationContext)
    {
        ApplicationContext = applicationContext;
    }

    private ApplicationContext ApplicationContext { get; set; }
    private NVList _nvList;

    public async Task<NVList> GetNVList()
    {
        if (_nvList == null)
        {
            var dp = ApplicationContext.GetRequiredService<IDataPortal<NVList>>();
            _nvList = await dp.FetchAsync();
        }
        return _nvList;
    }
}
```

Using this pattern, any time you need access to cached objects, you inject `CacheManager` into your page or whatever, and call the appropriate Get method to get the object you wanted to be cached.

**Reply by CROprogrammer** [reply]
I've written my CacheManager class, and it uses generics
`public class CacheManager<NVList, K, V> where NVList : NameValueListBase<K, V>`

and now I want to use it in my WPF page, something like this:
`<csla:CslaDataProvider x:Key="RoleList"
      ObjectType="{x:Type proj:CacheManager}"
      FactoryMethod="GetNVList"
      IsAsynchronous="True" />`

But I do not know how to use it in WPF when I need to declare generic arguement types of CacheManager. Any idea how to solve it?

**Reply by rockfordlhotka** [reply]
In `Program.cs` when you register your generic type with dependency injection, just register the generic:

```c#
  services.AddScoped<CacheManager<>>();
```

This allows DI to resolve the service based on the type parameters you use when requesting the instance.

If you are using WPF with DI to create each page, then you'd request the type in the page's constructor:

```c#
  public MyPage<CacheManager<RoleList, string, string> cache)
  { }
```

If you are using a centralized static method to access `ApplicationContext` then you'd probably request the type in the page's loaded event or something like that:

```c#
  private void Page_Loaded()
  {
    var cache = App.ApplicationContext.GetRequiredService(typeof(CacheManager<RoleList, string, string));
  }
```
